### PR TITLE
Overflow audits

### PR DESCRIPTION
**Pull request by DemiMarie-parity**
This adds checks for overflow in some places.  In others, it flags (using words like ERROR or WARN) where more attention is needed.

### CODE REVIEW

File: frame/assets/src/lib.rs [hunk 1]
```diff
@@ -179,7 +179,9 @@ decl_module! {
 			ensure!(origin_balance >= amount, "origin account balance must be greater than or equal to the transfer amount");
 
 			Self::deposit_event(RawEvent::Transferred(id, origin, target.clone(), amount));
+			// we check above that this will not overflow
 			<Balances<T>>::insert(origin_account, origin_balance - amount);
+			// sum of all issuances cannot exceed u64, so this will not overflow WARN
```
Comment (gavofyork) [inline frame/assets/src/lib.rs]: this is not true, as i said in the chat. however, *in the polkadot codebase* `TotalIssuance` will make it above 2**96 and `Balance` is `u128`.

File: frame/authorship/src/lib.rs [hunk 2]
```diff
@@ -238,6 +238,7 @@ impl<T: Trait> Module<T> {
 
 			T::EventHandler::note_uncle(
 				author.clone().unwrap_or_default(),
+				// WARN `uncle` cannot be after `now`
```
Comment (gavofyork) [inline frame/authorship/src/lib.rs]: why is this true?

Comment (gavofyork) [reply frame/authorship/src/lib.rs]: ok - i guess these are *our* uncles.

File: frame/balances/src/lib.rs [hunk 3]
```diff
@@ -851,6 +864,7 @@ where
 	fn issue(mut amount: Self::Balance) -> Self::NegativeImbalance {
 		<TotalIssuance<T, I>>::mutate(|issued|
 			*issued = issued.checked_add(&amount).unwrap_or_else(|| {
+				// cannot overflow: max_value >= any Self::Balance
```
Comment (gavofyork) [inline frame/balances/src/lib.rs]: tautologies like this comment don't need to be in the codebase.

File: frame/balances/src/lib.rs [hunk 4]
```diff
@@ -819,6 +831,7 @@ where
 	type NegativeImbalance = NegativeImbalance<T, I>;
 
 	fn total_balance(who: &T::AccountId) -> Self::Balance {
+		// WARN cannot overflow: total balance does not exceed u64
```
Comment (gavofyork) [inline frame/balances/src/lib.rs]: incorrect.

File: frame/authorship/src/lib.rs (hunk 2 above)
Comment (DemiMarie-parity) [reply frame/authorship/src/lib.rs]: @gavofyork I was not sure.  That is why I included this message!

Comment (DemiMarie-parity) [reply frame/authorship/src/lib.rs]: I added an explicit check for this.

**Comment by kianenigma** [comment]
What is the status here? 

if done, it would be nice to get it in for 2.0 @gnunicorn

**Comment by gnunicorn** [comment]
needs a merge first.

**Comment by gnunicorn** [comment]
someone needs to pick this up,  merge and resolve the conflicts first.

**Comment by bkchr** [comment]
@DemiMarie-parity ?

**Comment by DemiMarie-parity** [comment]
@bkchr will do

**Comment by bkchr** [comment]
Needs a merge.

**Comment by gnunicorn** [comment]
converted to draft as this is still in progress.

**Pull request closed without merging**